CRM/Activity/Selector/Search.php

```diff
@@ -174,16 +174,24 @@ public function __construct(
     $this->_activityClause = $activityClause;
 
     // CRM-12675
-    if (! CRM_Core_Permission::check('access CiviContribute')) {
-      $componentRestriction = ' (activity_type.component_id IS NULL OR activity_type.component_id <> 2) ';
-      if (empty($this->_activityClause)) {
-        $this->_activityClause = $componentRestriction;
-      }
-      else {
-        $this->_activityClause .= ' AND ' . $componentRestriction;
+    $components = CRM_Core_Component::getNames();
+    foreach ($components as $componentID => $componentName) {
+      if (!CRM_Core_Permission::check("access $componentName")) {
+        $componentClause[] = " (activity_type.component_id IS NULL OR activity_type.component_id <> {$componentID}) ";
```

Comment:
Would be nice to declare this variable before appending to it. I don't think php will complain about appending to an undefined variable, but IDEs will.